src/Microsoft.XmlSerializer.Generator/src/Sgen.cs

```diff
@@ -467,5 +475,77 @@ private static string GetTempAssemblyName(AssemblyName parent, string ns)
         {
             return parent.Name + ".XmlSerializers" + (ns == null || ns.Length == 0 ? "" : "." + ns.GetHashCode());
         }
+
+        private static void ParseReferences(string value, Dictionary<string, string> dictionary)
+        {
+            List<string> list = new List<string>();
+            if (value.Length > 0)
+            {
+                string[] entries = value.Split(new char[] { ';' });
+                for (int i = 0; i < entries.Length; i++)
+                {
+                    string entry = entries[i].Trim();
+                    if (entry == null || entry.Length == 0)
+                        continue;
+                    list.Add(entry);
+                }
+            }
+
+            foreach (var reference in list)
+            {
+                if (reference.EndsWith(".dll") || reference.EndsWith(".exe"))
+                {
+                    if (File.Exists(reference))
+                    {
+                        string filename = Path.GetFileNameWithoutExtension(reference);
+                        if (!string.IsNullOrEmpty(filename))
+                        {
+                            dictionary.Add(filename, reference);
+                        }
+                    }
+                }
+
+            }
+        }
+
+        private static Assembly SgenAssemblyResolver(object source, ResolveEventArgs e)
```

Comment:
This can be re-written without the need of the Dictionary or calling ParseReferences (unless you need the dictionary or that method for some other purpose).

```C#
    private static Assembly SgenAssemblyResolver(object source, ResolveEventArgs e)
    {
        try
        {
            if (string.IsNullOrEmpty(references) || string.IsNullOrEmpty(e.Name))
                return null;
            
            string[] eventArgs = e.Name.Split(',');
            
            if (eventArgs.Length == 0)
                return null;
            
            ReadOnlySpan<char> assemblyname = eventArgs[0].AsSpan();
            if (assemblyname.IsEmpty)
                return null;
            
            string[] entries = references.Split(';');
            for (int i = 0; i < entries.Length; i++)
            {
                ReadOnlySpan<char> reference = entries[i].AsSpan().Trim();
                if (reference.Length == 0)
                    continue;

                // need to add File.Exists overload that takes ReadOnlySpan<char>
                if ((reference.EndsWith(".dll") || reference.EndsWith(".exe"))
                    && File.Exists(reference))
                {
                    ReadOnlySpan<char> filename = Path.GetFileNameWithoutExtension(reference);
                    if (filename.Length != 0 && filename == assemblyname)
                        return Assembly.LoadFrom(reference);
                }
            }
        }
        catch (Exception exp)
        {
            if (exp is ThreadAbortException || exp is StackOverflowException || exp is OutOfMemoryException)
            {
                throw;
            }

            WriteWarning(exp, true);
        }

        return null;
    }
```

Reply:
Note, I didn't test the code above so you might need to make some tweaks.

Reply:
There is no point in making this code more complex by using Span. It is not performance critical and some extra allocations are not a problem.

Reply:
Good point. I think even outside of Span, there is some redundant use of Lists/etc that are not necessary and can be fixed. The same goes with creating a dictionary.

Reply:
Your check of `filename == assemblyname` will fail as the == operator on ReadOnlySpan<T> only checks if they point to the same memory and have the same length. See [the implementation of operator==](http://source.dot.net/#System.Private.CoreLib/shared/System/ReadOnlySpan.Fast.cs,196) for details.

---
In reply to: [170434450](https://github.com/dotnet/corefx/pull/27425#discussion_r170434450) [](ancestors = 170434450)

Reply:
Yes, of course. It should be SequenceEqual.